File: app/src/main/java/me/chrislane/accudrop/fragment/ReplaySideViewFragment.java
package me.chrislane.accudrop.fragment;

import android.app.Activity;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.ColorFilter;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.PixelFormat;
import android.graphics.PointF;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.widget.ImageView;

import java.util.List;

import me.chrislane.accudrop.R;
import me.chrislane.accudrop.presenter.ReplaySideViewPresenter;

public class ReplaySideViewFragment extends Fragment {

    private static final String TAG = ReplaySideViewFragment.class.getSimpleName();
    private ReplaySideViewPresenter presenter;
    private List<List<PointF>> screenPointsList;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        presenter = new ReplaySideViewPresenter(this);
    }

    /**
     * Redraw the side view lines.
     */
    public void updateDrawable(boolean generatePoints) {
        Activity activity = requireActivity();
        SideViewDrawable drawable = new SideViewDrawable(generatePoints);
        ImageView drawBox = activity.findViewById(R.id.replay_draw_area);
        drawBox.setImageDrawable(drawable);
    }

    public void setScreenPointsList(List<List<PointF>> screenPoints) {
        this.screenPointsList = screenPoints;
        updateDrawable(false);
    }

    /**
     * This drawable creates lines between points generates by the <code>ReplaySideViewPresenter</code>.
     */
    private class SideViewDrawable extends Drawable {

        private final boolean generatePoints;

        public SideViewDrawable(boolean generatePoints) {
            this.generatePoints = generatePoints;
        }

        @Override
        public void draw(@NonNull Canvas canvas) {
            // Paint the ground
            Paint paint = new Paint();
            paint.setColor(Color.GREEN);
            canvas.drawRect(0, (float) (canvas.getHeight() * 0.95), canvas.getWidth(), canvas.getHeight(), paint);

            if (generatePoints) {
                presenter.produceViewPositions(canvas.getWidth(), canvas.getHeight(), 20);
            }

            if (screenPointsList != null) {
                paint.setColor(Color.RED);
                paint.setStyle(Paint.Style.STROKE);
                paint.setStrokeWidth(10);

                Path path = new Path();
                for (List<PointF> screenPoints : screenPointsList) {
                    if (!screenPoints.isEmpty()) {
                        path.moveTo(screenPoints.get(0).x, canvas.getHeight() - screenPoints.get(0).y);
                        for (PointF point : screenPoints) {
                            path.lineTo(point.x, canvas.getHeight() - point.y);
                        }
                        canvas.drawPath(path, paint);
                    }
                }
            }
        }

        @Override
        public void setAlpha(int i) {

        }

        @Override
        public void setColorFilter(@Nullable ColorFilter colorFilter) {

        }

        @Override
        public int getOpacity() {
            return PixelFormat.UNKNOWN;
        }
    }
}

File: app/src/main/java/me/chrislane/accudrop/fragment/ReplaySideViewFragment.kt
package me.chrislane.accudrop.fragment

import android.graphics.*
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.support.v4.app.Fragment
import android.widget.ImageView
import me.chrislane.accudrop.R
import me.chrislane.accudrop.presenter.ReplaySideViewPresenter

class ReplaySideViewFragment : Fragment() {
    private lateinit var presenter: ReplaySideViewPresenter
    private var screenPointsList: MutableList<MutableList<PointF>>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        presenter = ReplaySideViewPresenter(this)
    }

    /**
     * Redraw the side view lines.
     */
    fun updateDrawable(generatePoints: Boolean) {
        val activity = requireActivity()
        val drawable = SideViewDrawable(generatePoints)
        val drawBox = activity.findViewById<ImageView>(R.id.replay_draw_area)
        drawBox.setImageDrawable(drawable)
    }

    fun setScreenPointsList(screenPoints: MutableList<MutableList<PointF>>) {
        this.screenPointsList = screenPoints
        updateDrawable(false)
    }

    /**
     * This drawable creates lines between points generates by the `ReplaySideViewPresenter`.
     */
    private inner class SideViewDrawable(private val generatePoints: Boolean) : Drawable() {

        override fun draw(canvas: Canvas) {
            // Paint the ground
            val paint = Paint()
            paint.color = Color.GREEN
            canvas.drawRect(0f, (canvas.height * 0.95).toFloat(), canvas.width.toFloat(), canvas.height.toFloat(), paint)

            if (generatePoints) {
                presenter.produceViewPositions(canvas.width, canvas.height, 20)
            }

            if (screenPointsList != null) {
                paint.color = Color.RED
                paint.style = Paint.Style.STROKE
                paint.strokeWidth = 10f

                val path = Path()
                for (screenPoints in screenPointsList!!) {
                    if (!screenPoints.isEmpty()) {
                        path.moveTo(screenPoints[0].x, canvas.height - screenPoints[0].y)
                        for (point in screenPoints) {
                            path.lineTo(point.x, canvas.height - point.y)
                        }
                        canvas.drawPath(path, paint)
                    }
                }
            }
        }

        override fun setAlpha(i: Int) {

        }

        override fun setColorFilter(colorFilter: ColorFilter?) {

        }

        override fun getOpacity(): Int {
            return PixelFormat.UNKNOWN
        }
    }

    companion object {
        private val TAG: String = ReplaySideViewFragment::class.java.simpleName
    }
}
